src/Data/Dwarf.hs

```diff
@@ -430,90 +417,139 @@ getDieAndSiblings cuContext = do
+    DW_FORM_ref1 -> RealizedAttr . DW_ATVAL_REF . inCU cu <$> getWord8
+    DW_FORM_ref2 -> RealizedAttr . DW_ATVAL_REF . inCU cu <$> derGetW16 end
+    DW_FORM_ref4 -> RealizedAttr . DW_ATVAL_REF . inCU cu <$> derGetW32 end
+    DW_FORM_ref8 -> RealizedAttr . DW_ATVAL_REF . inCU cu <$> derGetW64 end
+    DW_FORM_ref_udata ->  RealizedAttr . DW_ATVAL_REF . inCU cu <$> getULEB128
     DW_FORM_indirect -> getForm cuContext . DW_FORM =<< getULEB128
-    -- new in Dwarf version 4
-    DW_FORM_sec_offset -> DW_ATVAL_UINT <$> desrGetOffset end enc
-    DW_FORM_exprloc -> DW_ATVAL_BLOB <$> getByteStringLen getULEB128
-    DW_FORM_flag_present -> pure $ DW_ATVAL_BOOL True
-    DW_FORM_ref_sig8 -> DW_ATVAL_UINT . fromIntegral <$> derGetW64 end
-    DW_FORM_strx -> unimplForm form
-    DW_FORM_addrx -> unimplForm form
-    DW_FORM_ref_sup4 -> unimplForm form
-    DW_FORM_strp_sup -> unimplForm form
-    DW_FORM_data16 -> unimplForm form
-    DW_FORM_line_strp -> unimplForm form
-    DW_FORM_implicit_const -> unimplForm form
-    DW_FORM_loclistx -> unimplForm form
-    DW_FORM_rnglistx -> undefined
-    DW_FORM_ref_sup8 -> undefined
-    DW_FORM_strx1 -> unimplForm form
-    DW_FORM_strx2 -> unimplForm form
-    DW_FORM_strx3 -> unimplForm form
-    DW_FORM_strx4 -> unimplForm form
-    DW_FORM_addrx1 -> unimplForm form
-    DW_FORM_addrx2 -> unimplForm form
-    DW_FORM_addrx3 -> unimplForm form
-    DW_FORM_addrx4 -> unimplForm form
-    _ -> unimplForm form
+    _ -> RealizedAttr <$> getEvaluableForm dc end enc tgt form
+  where
+      -- shared structure between addrx and strx operations where an offset is used to index an array and that value is an offset into a further table
+      -- takes a parser for the offset, a function that consumes the offset and produces the final attribute values and a way to retrieve the base offset
+      -- for the target table
+      parseIntegralOff :: (MonadFail m, Show a, Integral a) => Sections -> Get a -> (Sections -> SectionOffset-> m DW_ATVAL) -> (CUContext -> Maybe Word64) -> Word64 -> Get (ParsedForm m)
+      parseIntegralOff sections getter f getBase addrSize =
+        do
+          index <- getter
+          let off = (fromIntegral index) * addrSize
+          pure $ DelayedAttr (\ctx ->
+              let base = fromMaybe 0 (getBase ctx)
+                  secOff = SectionOffset (fromIntegral off + base) in
+                  f sections secOff)
+
+orElse :: Maybe a -> Maybe a -> Maybe a
+orElse left right = maybe left Just right
+
+unpackDelayed :: (Monad m) => CUContext -> ParsedForm m -> m DW_ATVAL
+unpackDelayed ctx (DelayedAttr f) = f ctx
+unpackDelayed _ (RealizedAttr attr) = pure attr
+
+-- TODO: I really wish we had lenses here
+-- | Update a CUContext field if the attributes are equal
+updateFld :: (CUContext -> Maybe Word64) -> (CUContext -> Maybe Word64 -> CUContext) -> DW_AT -> DW_AT ->  ParsedForm (StateT CUContext Get) -> StateT CUContext Get ()
```

Comment:
Another particularly long type signature here, which I think would benefit from being split up into multiple lines (with Haddock comments for each function argument).